### PR TITLE
2 1 validation

### PR DESCRIPTION
**Pull request by Miishale**
Frontend + backend validation
Tests zu backend validation
Änderung naming Link zu Url

### CODE REVIEW

File: src/main/java/ch/bfh/bti7081/view/NewSeminarView.java [hunk 1]
```diff
@@ -1,22 +1,183 @@
...
+        //Category-Settings
+        seminarCategory.setLabel("Kategorie");
+        seminarCategory.setItems(SeminarCategoryManager.getSeminarCategories());
+        seminarCategory.setTextRenderer(SeminarCategory::getName);
+        seminarCategory.setItemLabelGenerator(SeminarCategory::getName);
+        seminarCategory.setEmptySelectionAllowed(false);
 
+        //Binder-Configuration
+        binder.forField(seminarTitle).asRequired("Bitte Titel angeben").
+                withValidator(title -> title.trim().length() >= SeminarManager.minTitleLength, "Titel muss aus mind. "+SeminarManager.minTitleLength + " Zeichen bestehen").
```
Comment (stormrider959) [inline src/main/java/ch/bfh/bti7081/view/NewSeminarView.java]: gemäss MVP-Pattern sollte die View nicht direkt mit dem SeminarManager zu tun haben

**Review by stormrider959 — APPROVED** [review]
habe einen Kommentar gemacht zum MVP-Pattern, aber das schauen wir morgen mit dem Rest an.
bis dahin GO

File: src/main/java/ch/bfh/bti7081/view/NewSeminarView.java (hunk 1 above)
Comment (Miishale) [reply src/main/java/ch/bfh/bti7081/view/NewSeminarView.java]: geit da nume um die werte vor minimum längi, die chame guet ine angeri klass uslagere, i ha eifach nid so genau gwüsst wo wiuse sowou d klass seminar wie oh d view klass brucht. aber luegemer das doch morn ah :)

**Review by oppli — APPROVED** [review]
Cool!

File: src/main/java/ch/bfh/bti7081/model/seminar/Seminar.java [hunk 2]
```diff
@@ -10,78 +12,122 @@
     private String title;
     private LocalDateTime date;
     private SeminarCategory category;
-    private String link;
+    private String url;
```
Comment (Azarny) [inline src/main/java/ch/bfh/bti7081/model/seminar/Seminar.java]: Warum das renaming? :o
Zudem gehört das Validieren eigentlich NICHT in die Entitäten, sondern in die Manager

File: src/main/test/SeminarValidationTest.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        Assert.assertFalse(testSeminar.setUrl("www.göögle.com"));
+    }
+
+
+    //test for description
+    @Test
+    public void testDescriptionOK(){
+        Assert.assertTrue(testSeminar.setDescription("Description mit Ümläütén"));
+        Assert.assertTrue(testSeminar.setDescription("Description mit Z4hl3n"));
+        Assert.assertTrue(testSeminar.setDescription("Dies ist ein sehr sehr langer Descriptionnname denn es muss getestet werden ob auch sehr sehr lange Descriptionnnamen funktionieren"));
```
Comment (Azarny) [inline src/main/test/SeminarValidationTest.java]: Detail: Zeilenumbruch ;)

File: src/main/test/SeminarValidationTest.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+    @Test
+    public void testCategoryNOK(){
+        Assert.assertFalse(testSeminar.setCategory(null));
+    }
+
+    //test for url
+    @Test
+    public void testUrlOK(){
+        Assert.assertTrue(testSeminar.setUrl("https://docs.jboss.org/hibernate/beanvalidation/spec/2.0/api/javax/validation/constraints/NotBlank.html"));
```
Comment (Azarny) [inline src/main/test/SeminarValidationTest.java]: Detail: Zeilenumbruch ;)

File: src/main/test/SeminarValidationTest.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        Assert.assertFalse(testSeminar.setLocation("A"));
+    }
+
+
+    //test for title
+    @Test
+    public void testTitleOK(){
+        Assert.assertTrue(testSeminar.setTitle("Title mit Ümläütén"));
+        Assert.assertTrue(testSeminar.setTitle("Title mit Z4hl3n"));
+        Assert.assertTrue(testSeminar.setTitle("Dies ist ein sehr sehr langer Titlenname denn es muss getestet werden ob auch sehr sehr lange Titlennamen funktionieren"));
```
Comment (Azarny) [inline src/main/test/SeminarValidationTest.java]: Detail: Zeilenumbruch :)

File: src/main/test/SeminarValidationTest.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        Assert.assertFalse(testSeminar.setPlz(1221314115));testSeminar.setPlz(0);
+    }
+
+
+    //test for location
+    @Test
+    public void testLocationOK(){
+        Assert.assertTrue(testSeminar.setLocation("Location mit Ümläütén"));
+        Assert.assertTrue(testSeminar.setLocation("Locati0n mit Z4hl3n"));
+        Assert.assertTrue(testSeminar.setLocation("Dies ist ein sehr sehr langer Locationnname denn es muss getestet werden ob auch sehr sehr lange Locationnnamen funktionieren"));
```
Comment (Azarny) [inline src/main/test/SeminarValidationTest.java]: Detail: Zeilenumbruch :)
Zudem... Wollen wir wirklich so lange Locations zulassen? Wenn ja, warum?

**Review by Azarny — CHANGES_REQUESTED** [review]

File: src/main/java/ch/bfh/bti7081/model/seminar/Seminar.java [hunk 7]
```diff
@@ -10,78 +12,122 @@
...
     private String description;
 
+
+
     public String getStreet() {
         return street;
     }
 
-    public void setStreet(String street) {
+    public boolean setStreet(String street) {
```
Comment (Azarny) [inline src/main/java/ch/bfh/bti7081/model/seminar/Seminar.java]: Die serverseitige Validierung gehört eigentlich nicht in unsere Entitäten, sondern in die Manager...

File: src/main/java/ch/bfh/bti7081/presenter/NewSeminarPresenter.java [hunk 8]
```diff
@@ -32,7 +32,7 @@ private Seminar convertDTOtoModel(NewSeminarDTO frontendObject) throws Exception
 
         modelObject.setTitle(frontendObject.getTitle());
         modelObject.setDescription(frontendObject.getDescription());
-        modelObject.setLink(frontendObject.getLink());
+        modelObject.setUrl(frontendObject.getLink());
```
Comment (yaronwalter7) [inline src/main/java/ch/bfh/bti7081/presenter/NewSeminarPresenter.java]: Finde die Umbenennung positiv.

File: src/main/java/ch/bfh/bti7081/view/NewSeminarView.java (hunk 1 above)
Comment (yaronwalter7) [reply src/main/java/ch/bfh/bti7081/view/NewSeminarView.java]: Stimme in diesem Punkt nach der Diskussion mit dem Dozenten zu, m.E. muss der Aufruf an den Presenter gehen und der muss die Länge zurückgeben, o.ä.

File: src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java [hunk 9]
```diff
@@ -28,57 +37,103 @@ public void deleteSeminar(Integer id) {
         throw new IllegalArgumentException("Not implemented yet.");
     }
 
-    private static List<Seminar> mockSeminaries(){
+    public static String validateSeminar(Seminar seminar) {
```
Comment (yaronwalter7) [inline src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java]: Aus meiner Sicht ist die Verwendung eines Strings zur Ausgabe des Validierungsresultats nicht optimal. 
M.E. wäre es besser hier einen Boolean zurückzugeben. (true/false)
Das Backend wird ja nicht durch einen Menschen unterhalten, sondern man will beim Speichern m.E. allenfalls validateSeminar(seminar) aufrufen und den Wert gleich verwenden, ohne String-Comparison. 
Im Frontend sieht man diese Meldungen ja nicht... 
Allenfalls könnte man sich überlegen, beim false eine exception zu werden, da ein invalides Seminar im Backend m.E. einen Fehler im Frontend bedeutet und damit sowieso ein Handling notwendig ist...

**Review by yaronwalter7 — CHANGES_REQUESTED** [review]
Da auch noch nicht wirklich mergbar noch kein Approve...

File: src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java (hunk 9 above)
Comment (Azarny) [reply src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java]: Es kommt darauf an was man machen will. Wenn du hier einen String mit der Fehlermeldung zurückgibst wo du genau sagen kannst was falsch lief, ist das gut. Wenn du dann ein boolean willst kannst du prüfen ob der String leer ist.

File: .idea/runConfigurations/start_App.xml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 <component name="ProjectRunConfigurationManager">
   <configuration default="false" name="start App" type="SpringBootApplicationConfigurationType" factoryName="Spring Boot">
-    <module name="ch.bfh.bti7081.s2019.yellow" />
+    <module name="SEnD" />
```
Comment (Azarny) [inline .idea/runConfigurations/start_App.xml]: Wenn du es bei dir anpassen musst, commite das nicht, sonst funktioniert es bei allen anderen nicht mehr

File: src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java [hunk 11]
```diff
@@ -84,6 +94,52 @@ public Seminar createSeminar(Seminar seminar) {
         throw new IllegalArgumentException("Not implemented yet.");
     }
 
+        public static String validateSeminar(Seminar seminar) {
```
Comment (Azarny) [inline src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java]: statische Methoden sind nicht gut... Gibt es einen bestimmten Grund warum diese Methode statisch sein muss?

File: src/main/java/ch/bfh/bti7081/model/seminar/Seminar.java [hunk 12]
```diff
@@ -42,10 +42,10 @@
     @JoinColumn(name = PREFIX + "CATEGORY")
     private SeminarCategory category;
 
-    @Column(name = PREFIX + "URL", length = 255)
+    @Column(name = PREFIX + "URL")
```
Comment (Azarny) [inline src/main/java/ch/bfh/bti7081/model/seminar/Seminar.java]: why?

File: src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java [hunk 13]
```diff
@@ -14,6 +14,16 @@
     // manages the communication between backend and frontend
     // the list of methods isn't completed yet, just some sample methods for the class diagramm
 
+    // some variables for the validation, used in SeminarManager.java and Seminar.java
+    public static int MAXYEARSINFUTURE = 5;
```
Comment (Azarny) [inline src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java]: Falsches Format der Variablen. Wenn Variablen aus mehreren Wörtern bestehen, werden diese (wenn es Konstanten sind) mit einem _ getrennt. Beispiel hier: MAX_YEARS_IN_FUTURE

File: src/main/test/SeminarManagerTest.java [hunk 14]
```diff
@@ -131,4 +132,245 @@ public void keywordFilter() {
...
+        Assert.assertTrue(SeminarManager.validateSeminar(testSeminar).contains("house number"));
+        testSeminar.setHouseNumber("   ");
+        Assert.assertTrue(SeminarManager.validateSeminar(testSeminar).contains("house number"));
+    }
+
+    //test for PLZ
+    @Test
+    public void testPlzOK(){
+        testSeminar.setPlz(3132);
+        Assert.assertEquals("", SeminarManager.validateSeminar(testSeminar));
```
Comment (Azarny) [inline src/main/test/SeminarManagerTest.java]: Im OK Fall könnte man auch weniger Fälle prüfen, meistens prüft man 1-2 OK Fälle, oder einfach die Edge-Cases. Aber natürlich darf man auch mehr testen :)

**Review by Azarny — CHANGES_REQUESTED** [review]
Geht in die richtige Richtung :)

File: src/main/java/ch/bfh/bti7081/view/NewSeminarView.java (hunk 1 above)
Comment (Miishale) [reply src/main/java/ch/bfh/bti7081/view/NewSeminarView.java]: Umgesetzt in presenter und mit Enum (isch huere geil imfau bini richtig stouz)

File: src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java (hunk 11 above)
Comment (Miishale) [reply src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java]: War an Yarons erste Version angelehnt, habe es nun angepasst

File: src/main/java/ch/bfh/bti7081/model/seminar/Seminar.java (hunk 12 above)
Comment (Miishale) [reply src/main/java/ch/bfh/bti7081/model/seminar/Seminar.java]: Wiu z IntelliJ gmotzet het, heimer ja hüt besproche, isch wider drinn

File: src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java (hunk 13 above)
Comment (Miishale) [reply src/main/java/ch/bfh/bti7081/model/manager/SeminarManager.java]: umgesetzt (i mim krasse enum)

File: src/main/test/SeminarManagerTest.java (hunk 14 above)
Comment (Miishale) [reply src/main/test/SeminarManagerTest.java]: ich prüfe 1 OK fall und die 4 edge cases... (dank dem hani de btw e fähler gfunge, isch auso scho guet wies isch)

File: .idea/runConfigurations/start_App.xml (hunk 10 above)
Comment (Miishale) [reply .idea/runConfigurations/start_App.xml]: done

**Review by yaronwalter7 — APPROVED** [review]
Finde ich super. 
Rundum glücklich, I LIKE!

**Review by Azarny — APPROVED** [review]
